### PR TITLE
Added a function to analyze a live container

### PR DESCRIPTION
**Pull request by m1-key**
I have added a function named analyze_live in
tern/analyze/default/container/single_layer.py which
takes an image layer and return a list of all the packages
found in that layer.

Solves issue - #858
Work towards - #850

Signed-off-by: m1-key <shubhamtiwari.tiwari84@gmail.com>

### CODE REVIEW

**Comment by m1-key** [comment]
@nishakm I'm not sure whether I'm returning the correct thing or not, as here I'm returning a list of all packages.

File: tern/analyze/default/container/single_layer.py [hunk 1]
```diff
@@ -148,3 +148,42 @@ def analyze_first_layer(image_obj, master_list, redo):
...
+    if not binary:
+        logger.warning(errors.no_package_manager)
+        image_layer.origins.add_notice_to_origins(
+            origin_image_layer, Notice(errors.no_package_manager, 'warning'))
+    # set a possible OS
+    get_os_style(image_layer, binary)
+    # if there is a binary, extract packages
+    if shell and binary:
+        # mount the given image layer
+        mount_first_layer(image_layer)
```
Comment (nishakm) [inline tern/analyze/default/container/single_layer.py:179]: We don't need this

File: tern/analyze/default/container/single_layer.py [hunk 2]
```diff
@@ -148,3 +148,42 @@ def analyze_first_layer(image_obj, master_list, redo):
     for p in image_obj.layers[0].packages:
         master_list.append(p)
     return shell
+
+
+def analyze_live(image_layer):
+    # set up a notice origin for the given image layer
+    origin_image_layer = 'Layer {}'.format(image_layer.layer_index)
+    # check if the layer is empty
+    if com.is_empty_layer(image_layer):
```
Comment (nishakm) [inline tern/analyze/default/container/single_layer.py:157]: Will have to check if we can actually use all the common and default_common functions in the "live" use case. I will need to take a look further before I can approve this PR.

File: tern/analyze/default/container/single_layer.py [hunk 3]
```diff
@@ -148,3 +148,42 @@ def analyze_first_layer(image_obj, master_list, redo):
...
+        logger.warning(errors.empty_layer)
+        image_layer.origins.add_notice_to_origins(
+            origin_image_layer, Notice(errors.empty_layer, 'warning'))
+        return None
+    # find the shell from the given image layer
+    shell = dcom.get_shell(image_layer)
```
Comment (nishakm) [inline tern/analyze/default/container/single_layer.py:163]: Same as the `com.is_empty_layer` comment.

File: tern/analyze/default/container/single_layer.py [hunk 4]
```diff
@@ -148,3 +148,42 @@ def analyze_first_layer(image_obj, master_list, redo):
...
+    get_os_style(image_layer, binary)
+    # if there is a binary, extract packages
+    if shell and binary:
+        # mount the given image layer
+        mount_first_layer(image_layer)
+        # core default execution on the given image layer
+        core.execute_base(image_layer, shell, binary)
+        # unmount
+        rootfs.undo_mount()
+        rootfs.unmount_rootfs()
```
Comment (nishakm) [inline tern/analyze/default/container/single_layer.py:184]: We don't need this.

File: tern/analyze/default/container/single_layer.py [hunk 5]
```diff
@@ -148,3 +148,42 @@ def analyze_first_layer(image_obj, master_list, redo):
...
+    # populate the package list with all packages found in the image layer
+    package_list = []
+    for p in image_layer.packages:
+        package_list.append(p)
+    return package_list
```
Comment (nishakm) [inline tern/analyze/default/container/single_layer.py:189]: Hmm...now that I think about it, we still need to maintain a master list of packages for this to work...

**Review by nishakm — COMMENTED** [review]
@m1-key I'm going to hold off on giving a specific review as my description of this scenario is half baked. I'll be filing more issues, and you are welcome to take those on. But please stand by :)

File: tern/analyze/default/container/single_layer.py (hunk 5 above)
Comment (m1-key) [reply tern/analyze/default/container/single_layer.py:189]: > Hmm...now that I think about it, we still need to maintain a master list of packages for this to work...

So, we will get master_list as an argument or master_list will be created just like above one ?

**Comment by m1-key** [comment]
@nishakm I removed the mounting and unmounting stuffs that are not required.

**Comment by nishakm** [comment]
There's some research I did which upends my assumptions when I created this issue. Sorry, @m1-key I'll have to close this PR and update the issues.

**Pull request closed without merging**